code/network/multi_pxo.cpp

```diff
@@ -1585,34 +1587,71 @@ void multi_pxo_strip_space(char *string1,char *string2)
 	}
 }
 
-// fire up the given URL
-void multi_pxo_url(char * /*url*/)
+static int open_url(const char *url)
 {
-#if 0
-	// execute the shell command
-	int r = (int) ShellExecute(NULL, NOX("open"), url, NULL, NULL, SW_SHOW);
-	if (r < 32) {		
-		switch (r) {
-			case 0:	
-			case ERROR_BAD_FORMAT: 
-			case SE_ERR_ACCESSDENIED: 
-			case SE_ERR_ASSOCINCOMPLETE: 
-			case SE_ERR_DDEBUSY:
-			case SE_ERR_DDEFAIL:
-			case SE_ERR_DDETIMEOUT:
-			case SE_ERR_DLLNOTFOUND:
-			case SE_ERR_OOM:
-			case SE_ERR_SHARE:			
-			case SE_ERR_NOASSOC:
-			case ERROR_FILE_NOT_FOUND:
-			case ERROR_PATH_NOT_FOUND:
-				popup(PF_USE_AFFIRMATIVE_ICON | PF_TITLE_RED | PF_TITLE_BIG,1,POPUP_OK,XSTR("Warning\nCould not locate/launch default Internet Browser",943));
-				break;
-		}					
+#if defined(_WIN32) || defined(__APPLE__)
+	const char *open_cmd = "open";
+#else
+	const char *open_cmd = "xdg-open";
+#endif
+
+	char s_url[256];
+
+	// make sure it's a valid web uri
+	if ( !SDL_strncasecmp(url, "http://", 7) || !SDL_strncasecmp(url, "https://", 8) ) {
+		SDL_strlcpy(s_url, url, SDL_arraysize(s_url));
+	} else {
+		SDL_snprintf(s_url, SDL_arraysize(s_url), "http://%s", url);
+	}
+
+#ifdef _WIN32
+	int rval = (int) ShellExecuteA(NULL, open_cmd, s_url, NULL, NULL, SW_SHOWNORMAL);
+
+	if (rval <= 32) {
+		return -1;
 	}
+
+	return 0;
+#else
+	int statval = 0;
+	pid_t mpid = fork();
+
+	if (mpid < 0) {
+		// nothing, will return error
+	} else if (mpid == 0) {
+		int rv = 0;
+
+		rv = execlp(open_cmd, open_cmd, s_url, (char *)0);
+
+		exit(rv);
+	} else {
+		waitpid(mpid, &statval, 0);
+
+		if ( WIFEXITED(statval) ) {
+			if (WEXITSTATUS(statval) == 0) {
+				return 0;
+			} else {
+				return -1;
```

Comment:
Just as a note, I cannot tell at all what is happening with this section after the #else

Reply:
For unix systems you need to fork off a child process in which to execute an external command. Depending on the return value of `fork()/vfork()` it indicates an error, whether it's the child process, or whether it's the parent process. If it's the child process then it execs the external command and exits. If it's the parent then it waits for the child to exit and checks the exit status to see if there was an error. So the highlighted block is the parent checking the return value of the child and returning the appropriate response.

Reply:
All right.  I'll give others a chance to look at it, since I don't know if what you're describing is being done correctly or not.  If it got into the icculus.org version, it's probably fine.